backend/src/main/java/de/bund/digitalservice/ris/norms/utils/EidConsistencyGuardian.java

```diff
@@ -108,127 +90,56 @@ private static boolean isReferenceValid(
-          final String currentEid = childElement.getAttribute("eId");
-          final String correctEid =
-              determineCorrectEid(childTypeCount.get(childTagName), currentEid, parentEid);
-
-          if (!currentEid.equals(correctEid)) {
-            updateReferences(currentEid, correctEid, rootElement);
-            childElement.setAttribute("eId", correctEid);
-            updateElementEids(
-                childElement,
-                correctEid,
-                rootElement); // Recursively update child elements with corrected eId
-          } else {
-            updateElementEids(
-                childElement,
-                currentEid,
-                rootElement); // Recursively update child elements with old eId
-          }
-        } else {
-          updateElementEids(childElement, null, rootElement);
-        }
-      }
-    }
-  }
+  private static void updateElementEids(final Node node, Element rootElement) {
+    final Map<String, String> oldToNewEIdMap = new HashMap<>();
+    List<Node> nodesToUpdate = new ArrayList<>();
+    nodesToUpdate.add(node);
 
-  private static String determineCorrectEid(int currentIndex, String oldEid, String newParentEid) {
-    final StringBuilder newEid = new StringBuilder();
+    while (!nodesToUpdate.isEmpty()) {
+      var currentNode = nodesToUpdate.removeLast();
 
-    String currentElementId;
+      if (currentNode instanceof Element element) {
+        final var expectedEId = EId.forNode(currentNode);
+        final var currentEId = EId.fromNode(currentNode);
 
-    // Check first if eId contains parent eIds
-    if (oldEid.contains("_")) {
-      // 1. ParentEid is not the same
-      final int lastIndex = oldEid.lastIndexOf('_');
+        if (expectedEId != currentEId) {
+          if (expectedEId.isPresent()) {
+            element.setAttribute("eId", expectedEId.get().value());
+          }
 
-      final String oldParentEid = oldEid.substring(0, lastIndex);
-      if (!oldParentEid.equals(newParentEid)) {
-        newEid.append(newParentEid);
-      } else {
-        newEid.append(oldParentEid);
-      }
-      currentElementId = oldEid.substring(lastIndex);
-    } else {
-      currentElementId = oldEid;
-    }
+          if (currentEId.isPresent() && expectedEId.isPresent()) {
+            oldToNewEIdMap.put(currentEId.get().value(), expectedEId.get().value());
+          }
+        }
 
-    // 2. Index of current node id not correct
-    final String elementEidNoIndex = currentElementId.substring(0, currentElementId.indexOf("-"));
-    final String currentElementIndex =
-        currentElementId.substring(currentElementId.indexOf("-") + 1);
-    if (Integer.parseInt(currentElementIndex) != (currentIndex)) {
-      newEid.append(elementEidNoIndex);
-      newEid.append("-");
-      newEid.append(currentIndex);
-    } else {
-      newEid.append(currentElementId);
+        nodesToUpdate.addAll(NodeParser.nodeListToList(currentNode.getChildNodes()).reversed());
```

Comment:
Why reversing the child nodes? isn't the result of getChildNodes() already in the order meaning the first one returned is the first child of the node?

Reply:
Yes but (for some reason i dont remember) we iterate in a depth-first way (and therefore use a last-in-first-out strategy) so the first child needs to be added last

Reply:
yes, i couldn't see a reason agains doing it breath-first instead of depth-first so i changed that